src/workerd/api/pyodide/pyodide.c++

```diff
@@ -300,13 +300,75 @@ kj::Array<kj::String> ArtifactBundler::parsePythonScriptImports(kj::Array<kj::St
     }
   }
 
-  // XXX: jsg doesn't support kj::Vector return types, so this seems to be the only way to do this.
-  auto builder = kj::heapArrayBuilder<kj::String>(result.size());
-  for (auto i = 0; i < result.size(); i++) {
-    builder.add(kj::mv(result[i]));
+  return result.releaseAsArray();
+}
+
+kj::String importToModuleFilename(kj::StringPtr import) {
+  auto builder = kj::heapArrayBuilder<char>(import.size() + 3);
+  for (auto i = 0; i < import.size(); i++) {
+    if (import[i] == '.') {
+      builder.add('/');
+    } else {
+      builder.add(import[i]);
+    }
   }
 
-  return builder.finish();
+  builder.add('.');
+  builder.add('p');
+  builder.add('y');
+  return kj::heapString(builder.finish());
+}
+
+kj::Array<kj::String> ArtifactBundler::filterPythonScriptImports(
+    kj::HashSet<kj::String> locals, kj::Array<kj::String> imports) {
+  kj::HashSet<kj::StringPtr> snapshotImportsSet;
+  for (auto import: ArtifactBundler::getSnapshotImports()) {
+    snapshotImportsSet.insert(kj::mv(import));
+  }
+
+  kj::HashSet<kj::String> filteredImports;
+  for (auto& import: imports) {
+    if (filteredImports.contains(import)) {
+      continue;  // Skip duplicates.
+    }
+
+    auto moduleFilename = importToModuleFilename(import);
+    if (!locals.contains(moduleFilename) && import != "js" &&
+        !snapshotImportsSet.contains(import)) {
+      filteredImports.insert(kj::mv(import));
+    }
+  }
+
+  kj::Vector<kj::String> filteredImportsVec;
+  for (auto& import: filteredImports) {
+    filteredImportsVec.add(kj::mv(import));
+  }
+
+  return filteredImportsVec.releaseAsArray();
+}
+
+kj::Array<kj::String> ArtifactBundler::filterPythonScriptImportsJs(
+    kj::Array<kj::String> locals, kj::Array<kj::String> imports) {
+  kj::HashSet<kj::String> localsSet;
+  for (auto& local: locals) {
+    localsSet.insert(kj::mv(local));
+  }
+  return ArtifactBundler::filterPythonScriptImports(kj::mv(localsSet), kj::mv(imports));
+}
+
+kj::Array<kj::StringPtr> ArtifactBundler::getSnapshotImports() {
+  kj::StringPtr imports[] = {"_pyodide.docstring"_kj, "_pyodide._core_docs"_kj, "traceback"_kj,
```

Comment:
Afaik, this can be `constexpr`. I added this to KJ, but I wonder if we are using the latest version right now.